Watch downloads

Frando opened the pull request:
Companion PR to hyperspace-org/hyperspace#5, see there for details.

Frando committed:
Watch downloads

mafintosh commented:
Nice! Instead of adding two new methods couldn’t we trigger this when a new download listener has been added?

Frando committed:
Use event listener changes for (un)watching downloads.

Frando commented:
@mafintosh I removed the two methods and now it just works by registering an event listener for the `download` event. Also rebased both this and the two companion PRs (in hyperspace and rpc) on top of master.

mafintosh commented on index.js:
```diff
@@ -358,6 +363,18 @@ class RemoteHypercore extends Nanoresource {
     this._onwaits = new FreeMap(1)
 
+    // Track listeners for the download event, and enable/disable download watching.
+    this.on('newListener', (event, _listener) => {
```
_listener is unused?

mafintosh commented on index.js:
```diff
@@ -358,6 +363,18 @@ class RemoteHypercore extends Nanoresource {
+        this._watchDownloads()
+      }
+    })
+    this.on('removeListener', (event, _listener) => {
```
_listener is unused?

Frando replied:
fixed

Frando committed:
Remove unused argument

mafintosh commented on index.js:
```diff
@@ -568,6 +590,22 @@ class RemoteHypercore extends Nanoresource {
+    try {
+      if (!this.opened) await this.open()
+      if (this.closed) return
+      await this._client.hypercore.watchDownloads({ id: this._id })
```
watchDownloadsNoReply and no await

mafintosh commented on index.js:
```diff
@@ -568,6 +590,22 @@ class RemoteHypercore extends Nanoresource {
+    try {
+      if (!this.opened) await this.open()
+      if (this.closed) return
+      await this._client.hypercore.unwatchDownloads({ id: this._id })
```
unwatchDownloadsNoReply and no await

Frando committed:
Use noReply for (un)watching downloads
